### ISSUE TITLE
[TASK] Final report - Methodology

### ISSUE THREAD

**Issue by Jarrus00** [WSU-CPTS415-ParquetParkour/Amazon-CoPurchasing#37]
_Background_:
We need a segment which describes how we are going about analyzing this dataset.

_Problem_:
How are we examining this dataset?  What did we expect to find?

_Success Criteria_:
1. From the prompt on Canvas: "This is where you give a detailed description of your primary contribution. It is especially important that this part be clear and well written so that we can fully understand what you did."
2. Pseudocode for algorithms and descriptions are included in this section.
3. References are cited and included in a subsection at the end.

**Comment by catdennis** [comment]
Methodology:

We are examining this dataset using multiple methods. Our main three methods in examining this data is implementing collaborative filtering, a backward expanding search algorithm, and a bidirectional search algorithm. Unfortunately due to complications we later discovered, we were unable to implement the backward expanding search algorithm, and bidirectional search algorithm. We have still introduced it into our presentation in hopes to show that if given more time we would hopefully be able to problem solve and get it working and implemented into our UI.

Collaborative filtering (CF) is a technique used by recommender systems.
collaborative filtering is a method of making automatic predictions (filtering) about the interests of a user by collecting preferences or taste information from many users (collaborating). The underlying assumption of the collaborative filtering approach is that if a person A has the same opinion as a person B on an issue, A is more likely to have B's opinion on a different issue than that of a randomly chosen person. With this in mind, we hope to be able to recommend similar products for users based on what other users rated it. Hopefully this will lead to an increase in both customer satisfaction and sales for amazon.
The implementation of this algorithm can be broken down into to two steps[1]:
1.	Look for users who share the same rating patterns with the active user (the user whom the prediction is for).
2.	Use the ratings from those like-minded users found in step 1 to calculate a prediction for the active user
3.	Build an item-item matrix determining relationships between pairs of items
4.	Infer the tastes of the current user by examining the matrix and matching that user's data

The pseudo code provided below used in the UI outputs recommended products based on the SQL query that outputs a group of customers who give a specific rating or purchase a specific category like books or movies.

Pseudo Code for collaborative filtering:

from sklearn.neighbors import NearestNeighbors

knn = NearestNeighbors(metric='cosine', algorithm='brute')
knn.fit(df.values)
distances, indices = knn.kneighbors(df.values, n_neighbors=3)
** The NearestNeighbors() in the sklearn.neighbors library can be used to calculate the distance ** between products using the cosine similarity and find the nearest neighbors for each **product.

def recommend_products(user, num_recommended_products):
    print('The list of the products {} Has purchased \n'.format(user))
    for m in df[df[user] > 0][user].index.tolist():
        print(m)
        print('\n')

    recommended_products = []
    for m in df[df[user] == 0].index.tolist():
        index_df = df.index.tolist().index(m)
        predicted_rating = df1.iloc[index_df, df1.columns.tolist().index(user)]
        recommended_products.append((m, predicted_rating))
*The code above is the function to show the recommended products for a selected user *using *the updated dataset.

sorted_rm = sorted(recommended_products, key=lambda x:x[1], reverse=True)
print('The list of the Recommended Products \n')
rank = 1
for recommended_products in sorted_rm[:num_recommended_products]:
    print('{}: {} - predicted rating:{}'.format(rank, recommended_product[0], recommended_product[1]))
    rank = rank + 1
*This portion of the code creates the print statement that gives the format for how the user *views their recommended products.

recommend_products('user_ID', 5)




























**Backwards expanding Search**
The backwards expanding search was a proposed algorithm following the investigation in to the BANKS system at http://www.cse.iitb.ac.in/banks/.

Backward expanding search is based on Dijkstra’s single source shortest path algorithm. Given a set of keywords, it ﬁrst ﬁnds, for each keyword term t, the set of nodes S that match t; the nodes in S are called keyword nodes[3]. Once a set of nodes S has been generated, each set get run through by Dijkstra’s algorithm in the reverse direction to find a common vertex. This vertex becomes the root with the leaves being any corresponding keyword nodes. This returns a connected tree of keyword nodes.

The main issue with this algorithm is the large overhead and exploring a large amount of unnecessary nodes. One major case that our data contained was the existence of large amount if frequent keywords, such as books or movie. This leads to a excessive amount of iterators that would be generated, creating too many iterations of Dijkstra’s algorithm to be running concurrently.

We failed to implement this algorithm into our system due to time constraints, the size of our database, and frequency of keywords in our database.

**Bidirectional search**
Bidirectional search is a graph searching algorithm which finds the smallest path from a source vertex to a goal vertex. It runs two simultaneous search –

-Forward search from source/initial vertex toward goal vertex
-Backward search from goal/target vertex toward source vertex

The idea of bidirectional search is to search forward from the start and backward from the goal simultaneously. When the two search frontiers intersect, the algorithm needs to construct a single path that extends from the start node through the frontier intersection to a goal node. It is a challenge to guarantee that the path found is optimal.[2]

In the article “Bidirectional expansion for keyword search on graph databases”, they merge all the iterators from the backward expanding search into a single iterator and a second iterator to go back on the explored nodes to check for a better path to all other keywords. The iterators go through two different priority queues that are chosen based of a spreading activation. The spreading activation is a way of creating the queues based on the amount a keyword appears in the nodes. 

We attempted to implement to an imitation of the algorithm in [2], but due to the difficulty of creating the two iterators and the limited amount time, failed to do so.









References –
-      “Collaborative Filtering  |  Machine Learning  |  Google Developers.” Google, Google, https://developers.google.com/machine-learning/recommendation/collaborative/basics.[1]
- 	V. Kacholia, S. Pandit, S. Chakrabarti, S. Sudarshan, R. Desai, and H. Karambelkar. Bidirectional expansion for keyword search on graph databases. In VLDB, 2005.[2]
-	G. Bhalotia, C. Nakhe, A. Hulgeri, S. Chakrabarti, and S. Sudarshan. Keyword searching and browsing in databases using BANKS. In ICDE, 2002[3]
-      J. Leskovec, L. Adamic and B. Adamic. [The Dynamics of Viral Marketing](http://www.cs.cmu.edu/~jure/pubs/viral-tweb.pdf). ACM Transactions on the Web (ACM TWEB), 1(1), 2007.[4]

**Comment by mccordsm** [comment]
modified back and bi sections

**Comment by Jarrus00** [comment]
Integrated into final report.

**Comment by catdennis** [comment]
Recommender systems are really critical in some industries as they can generate a huge amount of income when they are efficient. They have been known to companies as a competitive advantage and also be a way to stand out. This advantage comes from customers not wanting to browse the internet for hours trying find exactly what they are looking for but instead, having what they like come to them. As an example of how important these systems are, we can mention a real life scenario that happened a few years ago. Netflix organized a challenge called, the “Netflix prize” where the goal was to produce a recommender system that performs better than its own algorithm with a prize of 1 million dollars to win.
The main advantage of collaborative approaches is that they require no information about users like gender, age, location, or items like, realize year, author, etc. and, so, they can be used in many situations. However, as it only considers past interactions to make recommendations, collaborative filtering suffer from the “cold start problem”: it is impossible to recommend anything to new users or to recommend a new item to any user if either items have too few interactions to be efficiently handled.
